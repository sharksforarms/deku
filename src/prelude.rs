/*! Crate prelude

[What is a prelude?](std::prelude)
*/
pub use crate::error::{DekuError, NeedSize};
pub use crate::{
    deku_derive, reader::Reader, DekuContainerRead, DekuContainerWrite, DekuEnumExt, DekuRead,
    DekuReader, DekuUpdate, DekuWrite,
};
