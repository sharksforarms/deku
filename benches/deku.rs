use criterion::{black_box, criterion_group, criterion_main, Criterion};
use deku::prelude::*;

#[derive(Debug, PartialEq, DekuRead)]
struct DekuByte {
    data: u8
}

#[derive(Debug, PartialEq, DekuRead)]
#[deku(id_type = "u8")]
enum DekuEnum {
    #[deku(id = "0x01")]
    VariantA(u8),
}

#[derive(Debug, PartialEq, DekuRead)]
struct DekuVec {
    count: u8,
    #[deku(len = "count")]
    data: Vec<u8>
}

fn deku_read_byte(input: &[u8]) {
    let (_rest, v) = DekuByte::from_bytes((input, 0)).unwrap();
    assert_eq!(DekuByte { data: 0x01 }, v);
}

fn deku_read_enum(input: &[u8]) {
    let (_rest, v) = DekuEnum::from_bytes((input, 0)).unwrap();
    assert_eq!(DekuEnum::VariantA(0x02), v);
}

fn deku_read_vec(input: &[u8]) {
    let (_rest, v) = DekuVec::from_bytes((input, 0)).unwrap();
    assert_eq!(100, v.count);
    assert_eq!(100, v.data.len());
}

fn criterion_benchmark(c: &mut Criterion) {
    c.bench_function("deku_read_byte", |b| {
        b.iter(|| deku_read_byte(black_box([0x01].as_ref())))
    });

    c.bench_function("deku_read_enum", |b| {
        b.iter(|| deku_read_enum(black_box([0x01, 0x02].as_ref())))
    });

    let deku_read_vec_input = {
        let mut v = [0xFFu8; 101].to_vec();
        v[0] = 100u8;
        v
    };
    c.bench_function("deku_read_vec", |b| {
        b.iter(|| deku_read_vec(black_box(&deku_read_vec_input)))
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
