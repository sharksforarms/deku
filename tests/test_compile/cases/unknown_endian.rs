use deku::prelude::*;

#[derive(DekuRead)]
#[deku(endian = "variable")]
struct Test1 {
    a: u8,
}

#[derive(DekuRead)]
struct Test2 {
    #[deku(endian = "variable")]
    a: u8,
}

#[derive(DekuRead)]
#[deku(id_type = "u8", endian = "variable")]
enum Test3 {}

#[derive(DekuRead)]
#[deku(id_type = "u8")]
enum Test4 {
    #[deku(id = "1")]
    A(#[deku(endian = "variable")] u8),
}

fn main() {}
