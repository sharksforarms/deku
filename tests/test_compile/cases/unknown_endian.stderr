error[E0425]: cannot find value `variable` in this scope
 --> tests/test_compile/cases/unknown_endian.rs:3:10
  |
3 | #[derive(DekuRead)]
  |          ^^^^^^^^ not found in this scope
  |
  = note: this error originates in the derive macro `DekuRead` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find value `variable` in this scope
 --> tests/test_compile/cases/unknown_endian.rs:9:10
  |
9 | #[derive(DekuRead)]
  |          ^^^^^^^^ not found in this scope
  |
  = note: this error originates in the derive macro `DekuRead` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find value `variable` in this scope
  --> tests/test_compile/cases/unknown_endian.rs:15:10
   |
15 | #[derive(DekuRead)]
   |          ^^^^^^^^ not found in this scope
   |
   = note: this error originates in the derive macro `DekuRead` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0425]: cannot find value `variable` in this scope
  --> tests/test_compile/cases/unknown_endian.rs:19:10
   |
19 | #[derive(DekuRead)]
   |          ^^^^^^^^ not found in this scope
   |
   = note: this error originates in the derive macro `DekuRead` (in Nightly builds, run with -Z macro-backtrace for more info)

warning: unreachable expression
  --> tests/test_compile/cases/unknown_endian.rs:15:10
   |
15 | #[derive(DekuRead)]
   |          ^^^^^^^^
   |          |
   |          unreachable expression
   |          any code following this `match` expression is unreachable, as all arms diverge
   |
   = note: `#[warn(unreachable_code)]` on by default
   = note: this warning originates in the derive macro `DekuRead` (in Nightly builds, run with -Z macro-backtrace for more info)
